tests/unit/models/test_projects.py

```diff
@@ -1446,6 +1429,65 @@ def test_project_provenance_invalid(self, provenance, project_yaml_data):
         with pytest.raises(pydantic.ValidationError, match=error):
             Project.unmarshal(project_yaml_data(provenance=provenance))
 
+    @pytest.mark.parametrize(
+        "key",
+        [
+            "command",
+            "stop_command",
+            "post_stop_command",
+            "reload_command",
+            "bus_name",
+        ],
+    )
+    @pytest.mark.parametrize(
+        "value,bad_chars",
+        [
+            pytest.param(
+                "mkbird --chirps 5",
+                None,
+                id="good",
+            ),
+            pytest.param(
+                "mkbird --chirps=5",
+                ["="],
+                id="has_bad_char",
+            ),
+            pytest.param(
+                'mkbird --chirps=1337 --name="81U3J@Y"', ['"', "@", "="], id="many_bad"
+            ),
+        ],
+    )
+    def test_app_command_lexicon(
+        self, app_yaml_data, key: str, value: str, bad_chars: list[str] | None
+    ):
+        """Verify that invalid characters in command fields raise an error."""
+        command = {key: value}
+        data = app_yaml_data(**command)
+        doc_link = (
+            "https://snapcraft.io/docs/snapcraft-yaml-schema#p-21225-appsapp-name"
+        )
+
+        if bad_chars:
+            err_msg = textwrap.dedent(
+                f"""\
+                {key.replace('_', '-')}: App commands must consist only of alphanumeric characters, spaces, and the following characters:
+                / . _ # : $ -
+                The following characters were encountered, but not allowed: {', '.join(sorted(bad_chars))}
+                For more complete details about the `command` key, see: {doc_link}
+            """
+            )
+
+            with pytest.raises(pydantic.ValidationError) as val_err:
+                Project.unmarshal(data)
+
+            assert err_msg in str(val_err.value)
+
+        else:
```

Comment:
These are two separate tests. One is `test_app_command_lexicon_success` and the other is `test_app_command_lexicon_failure`.